stakewise_cli/storages/database.py

```diff
@@ -0,0 +1,174 @@
+from base64 import b64encode
+from functools import cached_property
+from typing import List, Set
+from urllib.parse import urlparse
+
+import click
+import psycopg2
+from Crypto.Cipher import AES
+from Crypto.Random import get_random_bytes
+from eth_typing import ChecksumAddress, HexStr
+from eth_utils import add_0x_prefix
+from psycopg2.extras import execute_values
+from py_ecc.bls import G2ProofOfPossession
+from web3 import Web3
+
+from stakewise_cli.eth1 import get_operator_deposit_data_ipfs_link
+from stakewise_cli.eth2 import get_mnemonic_signing_key
+from stakewise_cli.ipfs import ipfs_fetch
+from stakewise_cli.networks import NETWORKS
+from stakewise_cli.queries import get_stakewise_gql_client
+from stakewise_cli.settings import IS_LEGACY
+from stakewise_cli.typings import DatabaseKeyRecord
+
+CIPHER_KEY_LENGTH = 32
+
+
+class Database:
+    def __init__(
+        self,
+        db_url: str,
+        operator: ChecksumAddress,
+        network: str,
+        mnemonic: str,
+    ):
+        self.db_url = db_url
+        self.sw_gql_client = get_stakewise_gql_client(network)
+        self.network = network
+        self.mnemonic = mnemonic
+        self.max_keys_per_validator = NETWORKS[network]["MAX_KEYS_PER_VALIDATOR"]
+        self.operator_address = operator
+
+    @cached_property
+    def cipher_key(self) -> bytes:
+        return get_random_bytes(CIPHER_KEY_LENGTH)
+
+    @cached_property
+    def cipher_key_str(self) -> str:
+        return b64encode(self.cipher_key).decode("ascii")
+
+    @cached_property
+    def keys(self) -> List[DatabaseKeyRecord]:
+        """
+        Returns prepared database key records that are in the latest deposit data.
+        """
+        keystores: List[DatabaseKeyRecord] = []
+        keys_count = len(self.operator_deposit_data_public_keys)
+        index = 0
+
+        with click.progressbar(
+            length=keys_count,
+            label="Syncing deposit data keystores\t\t",
+            show_percent=False,
+            show_pos=True,
+        ) as bar:
+            while index < keys_count:
+                private_key = get_mnemonic_signing_key(self.mnemonic, index, IS_LEGACY)
+                public_key = Web3.toHex(G2ProofOfPossession.SkToPk(private_key.key))
+
+                public_key = add_0x_prefix(public_key)
+
+                if public_key not in self.operator_deposit_data_public_keys:
```

Comment:
It could be that the public key is already registered, but is not located in operator deposit data.
I suggest we include all the keystores that are either in the operator's deposit data or are already registered.